### ISSUE TITLE
feat: React Router v6 support

### ISSUE THREAD

**Issue by ptmkenny** [ionic-team/ionic-framework#24177]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already include this feature request, without success.

### Describe the Feature Request

Ionic React currently uses React Router 5, which is [tagged as classic on yarn](https://classic.yarnpkg.com/en/package/react-router-dom).

[React Router](https://reactrouter.com/) v6 received a stable release on [November 3, 2021](https://remix.run/blog/react-router-v6).  Three years passed and Ionic released Ionic React versions 6, 7, and 8, but none of these supported React Router 6.

It would be great to be able to use a version of React Router that is actively being updated and makes use of the new features in React 18 and 19.

### Describe the Use Case

React Router 6 adds features with React 18 and 19 that v5 does not.

### Describe Preferred Solution

`@ionic/react-router` should update to React Router 6.

### Issues Fixed in React Router 6 but not 5

React Router 5 is only getting critical security fixes since development has effectively ended.

* https://github.com/remix-run/react-router/issues/10814

**Comment by fortesp** [comment]
I also have this question. 
Now that we have Ionic version 6, i was trying to upgrade to react router version 6 as well, but it seems not yet possible.

I get the following error:

```
Creating an optimized production build...
Failed to compile.

./node_modules/@ionic/react-router/dist/index.esm.js
Attempted import error: 'withRouter' is not exported from 'react-router-dom'
```

**Comment by ptmkenny** [comment]
@fortesp This is a feature request, not a bug report.  Upgrading to react router v6 is not supported in Ionic now, and it will not be until the Ionic team adds code to support it; there's no "workaround" unless you want to write the code and contribute a PR yourself.

**Comment by fortesp** [comment]
@ptmkenny not reporting any bug, just enforcing the idea. Not sure i understand the rest of your comment. The idea is to understand when it will be supported.

**Comment by zhangriyueming** [comment]
Like to hear a timeline for this.

**Comment by jrobeson** [comment]
the ionic upgrade guide  mentions react-router v6 here (https://ionicframework.com/docs/intro/upgrading-to-ionic-6#react), so isn't this solved? Although there's no upgrade guide for the router yet.

**Comment by ptmkenny** [comment]
@jrobeson No, this is not solved.  The documentation refers to `@ionic/react-router`, which is Ionic's custom implementation of [react-router](https://www.npmjs.com/package/react-router).  `@ionic/react-router` v6 is based on `react-router` v5, and this issue is about updating `@ionic/react-router` to `react-router` v6.

**Comment by jrobeson** [comment]
Hmm, after looking around , it seems the instructions on that page are premature, since the the `@ionic/react-router` itself still depends on on a 5.x version.  How did those instructions make it through?

**Comment by jrobeson** [comment]
@ptmkenny : yeah i accidentally read it as `react-router`. my fault. sorry.

**Comment by RayZik** [comment]
Hi! Any updates?

**Comment by CooleyGit** [comment]
Is there any timeline on this?

**Comment by XueMeijing** [comment]
Hi! Any updates?

**Comment by christianmat** [comment]
Any updates?

**Comment by daniel-escoto** [comment]
Any updates??

**Comment by I-am-abdulazeez** [comment]
Any update, please??

**Comment by ptmkenny** [comment]
I know people are excited about this, but **please STOP asking for updates.**  Every time someone posts to this thread, everybody who has subscribed gets a message.

**If you want to get updates, press the subscribe button.**  You don't need to post a message to get updates; just hit the subscribe button.

**If you want to show support for this issue, [give my original post a thumbs up](https://github.com/ionic-team/ionic-framework/issues/24177#issue-1046428266).  That way you can show support without spamming everyone who has subscribed to the thread.

### Update

Disclaimer: I am not affiliated with Ionic; I'm just the person who posted this issue.

This issue is postponed until the `react-router` team finishes their backwards compatibility package.  As stated in the [React Router guide for upgrading from 5 to 6](https://reactrouter.com/docs/en/v6/upgrading/v5):

> We recommend waiting for the backwards compatibility package to be released before upgrading apps that have more than a few routes.

Ionic React has a large number of users, and there's a lot of changes between `react-router` 5 and 6.  At present, because the backward compatibility package isn't ready, **when you upgrade from react-router 5 to 6, you have to update all the routing in your app at once.**  This means the developer has to do a lot of work.

Ionic minor releases (6.1 to 6.2, for example) do not have major breaking changes.  `react-router` 5 to 6 is currently a huge breaking change, so it is unlikely to land in ionic until version 7 (this is my guess, not an official statement from the ionic team).

Now, this might change if `react-router` releases a very easy-to-use backwards compatibility package, but that hasn't happened yet.

If you want to see this happen faster, please consider stepping up, writing some code, and getting a PR started.  Or go help `react-router` finish up their backwards compatibility work.

I'll close by saying thank you to the Ionic and `react-router` teams, who have done fantastic work giving us great open source software.

**Comment by DercilioFontes** [comment]
@Mahdi-Soultana,

React Router v6 is not supported. You need to use v5.

**Comment by Brian-McBride** [comment]
I dropped this in the forums as well. 
Mainly I'm hoping to get some help to iron out the issues. 
It's "mostly" there, I just see a couple of areas where the animations are not right. 

I am pretty sure it is due to the `ViewStacks` that I didn't use. There might need to be some additional lifecycle stuff dealing with parent outlets and their visibility. Maybe someone on the Ionic team who wrote the `@ionic/react-router` can take a look into this updated StackManager. It is largely the same, other than using the elements from the `react-router-dom` outlet to render the current page and stashing a ref to the previous page that was rendered to handle an animation after the page reloads. 

https://www.npmjs.com/package/@anatine/ionic-react-router

https://github.com/anatine/ionic-modules

**Comment by liho00** [comment]
react router v6 is awesome, im using react router v6 instead of @ionic/react-router now, and build my own navigation transition.

**Comment by Brian-McBride** [comment]
> react router v6 is awesome, im using react router v6 instead of @ionic/react-router now, and build my own navigation transition. Hope @ionic/react-router can adopt react router v6 sooner

I am curious on your method on animating one view over another and how you are dealing with the stack to pop off views and all that. Also, are you using CSS animation directly or something like Framer Motion?

**Comment by kevinclarkadstech** [comment]
> I know people are excited about this, but **please STOP asking for updates.** Every time someone posts to this thread, everybody who has subscribed gets a message.
> 
> **If you want to get updates, press the subscribe button.** You don't need to post a message to get updates; just hit the subscribe button.
> 
> **If you want to show support for this issue, [give my original post a thumbs up](https://github.com/ionic-team/ionic-framework/issues/24177#issue-1046428266). That way you can show support without spamming everyone who has subscribed to the thread.
> 
> ### Update
> Disclaimer: I am not affiliated with Ionic; I'm just the person who posted this issue.
> 
> This issue is postponed until the `react-router` team finishes their backwards compatibility package. As stated in the [React Router guide for upgrading from 5 to 6](https://reactrouter.com/docs/en/v6/upgrading/v5):
> 
> > We recommend waiting for the backwards compatibility package to be released before upgrading apps that have more than a few routes.
> 
> Ionic React has a large number of users, and there's a lot of changes between `react-router` 5 and 6. At present, because the backward compatibility package isn't ready, **when you upgrade from react-router 5 to 6, you have to update all the routing in your app at once.** This means the developer has to do a lot of work.
> 
> Ionic minor releases (6.1 to 6.2, for example) do not have major breaking changes. `react-router` 5 to 6 is currently a huge breaking change, so it is unlikely to land in ionic until version 7 (this is my guess, not an official statement from the ionic team).
> 
> Now, this might change if `react-router` releases a very easy-to-use backwards compatibility package, but that hasn't happened yet.
> 
> If you want to see this happen faster, please consider stepping up, writing some code, and getting a PR started. Or go help `react-router` finish up their backwards compatibility work.
> 
> I'll close by saying thank you to the Ionic and `react-router` teams, who have done fantastic work giving us great open source software.

The point of asking for updates is not to spam people subscribed to the topic, but help show how this is a high demand topic that should receive focus.

And I will be the next to ask ...ANy updates?  I am about to file a bug because I cannot update.  If Ionic want to use the library routers for React, Vue, and Angular, they need to keep their wrapper package up to date with major versions of the library routers.  If it was really released in Nov last year, that should be enough time to make changes to upgrade ionic-react-router to v6.

**Comment by ptmkenny** [comment]
A member of the Ionic team[ posted an update on the status of React Router 6 support here.](https://forum.ionicframework.com/t/ionic-react-so-many-render-issues-related-to-routing/222301/14?u=ptmkenny)

If there is high demand, give it a thumbs up.  Don't spam.

**Comment by kevinclarkadstech** [comment]
> A member of the Ionic team[ posted an update on the status of React Router 6 support here.](https://forum.ionicframework.com/t/ionic-react-so-many-render-issues-related-to-routing/222301/14?u=ptmkenny)
> 
> 
> 
> If there is high demand, give it a thumbs up.  Don't spam.

If I hadn't "spammed", then I would not know about that link with the update, now would I? 🙄

**Comment by piotr-cz** [comment]
I think that new major ionic release [v7.0.0-beta.0](https://github.com/ionic-team/ionic-framework/releases/tag/v7.0.0-beta.0) is an opportunity to reevaluate react-router v6 support.

**Comment by ebk46** [comment]
What about https://www.npmjs.com/package/react-router-dom-v5-compat? Seems like a relatively straightforward way to introduce compatibility with react-router v6.

**Comment by michaelmota** [comment]
Any updates?

**Comment by Brian-McBride** [comment]
> Any updates?

I've given up on the hope of a modern React router. I think the Iconic team is comprised of Angular devs, and the React port is really just that - a wrapper around the web components designed for Angular. 

It isn't the same, but Framework7 has a pretty nice mobile router. It is completely custom, not based on React Router at all. But has a solid navigation stack. I found it to be a good option on my last project.

**Comment by salemlf** [comment]
> > Any updates?
> 
> I've given up on the hope of a modern React router. I think the Iconic team is comprised of Angular devs, and the React port is really just that - a wrapper around the web components designed for Angular. 
> 
> It isn't the same, but Framework7 has a pretty nice mobile router. It is completely custom, not based on React Router at all. But has a solid navigation stack. I found it to be a good option on my last project. 

Yeah, it's disappointing. I was looking into [Tanstack Router](https://tanstack.com/router/v1), but not ideal as it's still in beta. Also not sure if it'd be compatible with Ionic's Capacitor/Cordova functionality for iOS and Android

**Comment by sean-perkins** [comment]
@craigbehnke it is unlikely that that branch will be the intended changes for React Router v6 support. It was mainly discovery to identify the work effort, pain points and path that Ionic Framework can take to upgrading the package. 

I have not looked into the compatibility package, but that is another path worth exploring. 

Ionic's router implementation makes use of class components (mainly the `StackManager`), which need to be migrated to hooks before they can be used with React Router v6. Many of the other changes in v6 (internalizing the history API, etc.) were fairly straightforward to navigate and function at some capacity in my discovery work. 

Developers will need time to migrate their apps to React Router v6 once Ionic provides support. I think it is likely we will run both v5 and v6 support in parallel for a time, to help resolve any v6 specific challenges before making it the default version and deprecating v5 support. 

We plan to support v6, but we do not have any specific timeline to share at this time. This is our most upvoted and community followed feature request, so we are actively working to bring this to you all.

**Comment by FranciscoKloganB** [comment]
It's very sad that Ionic still hasn't catched up to react router 6. So many things are great about it, ranging from better developer experience to better user experience. Backwards compability is important, but starting a new ionic project and being tied to react router 5 kind of sucks. Ionic could introduce a breaking change. Companies don't need to bump if they don't feel ready to. :(

**Comment by Berkmann18** [comment]
@FranciscoKloganB I totally agree. And looking at https://forum.ionicframework.com/t/ionic-react-so-many-render-issues-related-to-routing/222301/20, it seems like it's been a Work In Progress since mid-April (so nearly 6 months ago).

**Comment by inouiw** [comment]
> @craigbehnke it is unlikely that that branch will be the intended changes for React Router v6 support. It was mainly discovery to identify the work effort, pain points and path that Ionic Framework can take to upgrading the package.
> 
> I have not looked into the compatibility package, but that is another path worth exploring.
> 
> Ionic's router implementation makes use of class components (mainly the `StackManager`), which need to be migrated to hooks before they can be used with React Router v6. Many of the other changes in v6 (internalizing the history API, etc.) were fairly straightforward to navigate and function at some capacity in my discovery work.
> 
> Developers will need time to migrate their apps to React Router v6 once Ionic provides support. I think it is likely we will run both v5 and v6 support in parallel for a time, to help resolve any v6 specific challenges before making it the default version and deprecating v5 support.
> 
> We plan to support v6, but we do not have any specific timeline to share at this time. This is our most upvoted and community followed feature request, so we are actively working to bring this to you all.

@sean-perkins Supporting two versions if the api and the behaior is not exactly the same will likely be quite a lot of more work than just supporting v6 and fixing any bugs that may surface.

**Comment by nadavhalfon** [comment]
We really appreciate your work 🙏🏻
If you can tell us a little bit about where are you standing regarding the development it'll be nice.
I couldn't find an active branch or progress, I'm sure you are working on this but just to understand and to light the situation 🙏🏻
Thanks !

**Comment by 020AlievIB** [comment]
Do Ionic 7 support React Router v6?

**Comment by nadavhalfon** [comment]
> ```
> Do Ionic 7 support React Router v6?
> ```

As for now, no

**Comment by nadavhalfon** [comment]
> @craigbehnke it is unlikely that that branch will be the intended changes for React Router v6 support. It was mainly discovery to identify the work effort, pain points and path that Ionic Framework can take to upgrading the package.
> 
> I have not looked into the compatibility package, but that is another path worth exploring.
> 
> Ionic's router implementation makes use of class components (mainly the `StackManager`), which need to be migrated to hooks before they can be used with React Router v6. Many of the other changes in v6 (internalizing the history API, etc.) were fairly straightforward to navigate and function at some capacity in my discovery work.
> 
> Developers will need time to migrate their apps to React Router v6 once Ionic provides support. I think it is likely we will run both v5 and v6 support in parallel for a time, to help resolve any v6 specific challenges before making it the default version and deprecating v5 support.
> 
> We plan to support v6, but we do not have any specific timeline to share at this time. This is our most upvoted and community followed feature request, so we are actively working to bring this to you all.

Hey team!
All of the most voted issues in react are related to the ionRouter (animation problems, white screens and etc.), some of them are related to fallback routes and some are not.
I assume that you plan to support it on the next major (aka. ionic 8) because of the migration process people will have to do.
Unrelated to react router 6 I really think that the ionic react router is the main blocker to use ionic react for production (at least in our case), it has many problems that could have been resolved anyway.
Because we don't know what is your progress or general schedule, we don't know if to take a look and help with something that you might already changed.
Let me be clear I am sure that you are working on it and plan to do those improvements and fixes, it's been almost 6 months since the last update, so I'll really appreciate if you can give us a small update (even that your'e stuck will be good so we won't develop any hopes for our production schedule)

Thank you!

**Comment by nkeysinstil** [comment]
This would be really useful as it has blocked us from using https://ionicframework.com/docs/api/tabs

**Comment by nadavhalfon** [comment]
Hey, you announced Ionic 8 and still remain silence regarding this issue, It'll be very nice from the Ionic team to shine some bright on this issue, 2 years passed, 2 majors, at least tell us if it's intended to be on ionic 8 or not, people (regarding me) really trying to be patient and wait but you are not answering since Aug 23'

Thank you

**Comment by Brian-McBride** [comment]
## Current State

About a year ago, I looked into adapting v6 to my Ionic toolkit and uploading a PR. 

The React Router team has really moved into an SSR paradigm, and their updated library removed some of the functionality that v5 had. This, combined with the Iconic team being Angular developers, is my guess that there is a complete lack of progress here. 

SSR routing and the trend in web development are entirely different from mobile stack and tab navigation. Most of these libraries rely on the browser's built-in history. Some new mobile navigation features are even coming built-in to the browser, making it even less urgent for these teams to invest time.

## Solutions

### 1. Create a universal Ionic Router

This is really what the Ionic team should do. They already built a stack/tab navigation for Angular. It works better in all regards. The best thing to do is optimize it so that the core route management is agostic to any UI platform. Just really a state machine for routing. Then, build interface layers for Angular, Vue, and React. The community could then expand upon that. 

### 2. Bail on Ionic for routing: Framework7

I have not used it in production. However, I will say that https://framework7.io/ had a much better router in place for React.
Of course, if you end up using that router, you could use the entire framework for UI. 

The guy who maintains it also has a tailwind version, https://konstaui.com/, which is pretty cool. However, the core Framework7 is more maintained. 

Keep in mind, Framework7 and similar libs work with Capictor just fine.

**Comment by dbousamra** [comment]
Any further updates?

**Comment by nadavhalfon** [comment]
@FranciscoKloganB  I agree with you
but what really is not making sense is why the ionic team doesn't say what's this issue status, I would really appreciate even if you say that you have no progress or no intentions for making this change so at least we know for sure it's wroth forking it
why is it so hard to give an update here in the comments?
thank you if you reply, just want some transparency 😁

**Comment by krishanmarco** [comment]
This is not a production solution but  it will work for not too complex web/mobile apps - We have, at least managed to prototype this but still have to deal with the nitty gritty part of actually synching up the browser history between react-router@5 and react-router@6 in a sane manner - If that's even possible.

Anyways

**Install react-router@5 alongside your current react-router@6**
package.json
```json
{
  "dependencies": {
    "react-router-v5": "npm:react-router@5.3.4",
    "react-router-dom-v5": "npm:react-router-dom@5.3.4",
  }
}
```

**Place all the ionic react dependencies in a subfolder of your project (Say `src/ionic`)**

then

**Use you bundler to ensure that all the `react-router-dom` and `react-router` imports coming from the `src/ionic` folder and ionic's dependencies point to `react-router-dom-v5` instead of `react-router-dom`**

We @closeio we use webpack - It looks like this:

```
 ...[
      { regex: /^react-router-dom$/, target: 'react-router-dom' },
      { regex: /^react-router$/, target: 'react-router' },
    ].map(({ regex, target }) => {
      return new NormalModuleReplacementPlugin(regex, function (resource) {
      
        // All the ionic dependencies should point to v5
        const mobileContextMatchers = [
          /.*\/@capacitor.*/gi,
          /.*\/@ionic.*/gi,
          /.*\/ionicons.*/gi,
          /.*\/\.pnpm\/react-router-dom@5\.3\.4_react.*/gi,
          // /.*\/react-router-dom.*5\.3\.4.*/gi,
          /.*\/mobile\/src.*/gi,
        ];

        const shouldUseRR5 = mobileContextMatchers.some((s) =>
          s.test(resource.context),
        );

        console.log(resource.context, `on ${target} shouldUseRR5=${shouldUseRR5}`);

        if (shouldUseRR5) {
          resource.request = `${target}-v5`;
        } else {
          resource.request = target;
        }
      });
    }),
```

We then rendered the React Router 6 `RouterProvider` with a child route. The child route imports and renders the `IonApp` which circles back and renders components from our actual app.

This works and I'll post again if we can get a decent enough compatibility layer between the two in a way that makes them place nice together.

**Comment by franelias** [comment]
Hi there! Any update on this?

**Comment by esmatbh** [comment]
Any movement on support for react-router v6? This issue has been open for years, and it’s blocking modern app development with Ionic and React. The lack of compatibility has left projects like ours forced to stay on react-router v5 or build tedious workarounds. Is there a planned fix or some guidance on where we stand with this?

**Comment by superm123** [comment]
I was forced to completely rip out this

Op Wo. 30 Okt. 2024 om 09:29 het esmatbh ***@***.***> geskryf:

**Comment by nadavhalfon** [comment]
Hey any update?

**Comment by brandyscarney** [comment]
Hello everyone,

I wanted to provide an update on this issue — apologies for not doing so sooner. This is our most upvoted and highest-priority issue, so it remains a key focus for us.

Upgrading React Router has been a highly complex task, which is why it has taken longer than we initially hoped. There have already been two attempts to address this upgrade. The [most recent attempt](https://github.com/ionic-team/ionic-framework/compare/main...sp/react-router-6-take-2) made significant progress, but there are still some unresolved issues.

Currently, we have an open task underway, and our team is actively investigating and addressing the remaining challenges.

Thank you for your patience and support as we work towards resolving this.

**Comment by babycourageous** [comment]
Thanks for the recap @brandyscarney !
I can only imagine the undertaking that upgrade path is. Appreciate all the work going into it. Good luck!

**Comment by piotr-cz** [comment]
Meanwhile React Router v7 came up:
https://remix.run/blog/react-router-v7

**Comment by ptmkenny** [comment]
@piotr-cz Thanks, I've taken the liberty of updating this feature request to target React Router 7, since it seems that React Router 6 will quickly be phased out in favor of React Router 7 (the docs say there are no breaking changes between React Router 6 and 7 if you having all future flags enabled).

**Comment by aeharding** [comment]
Open issue which is fixed in v6+, but not patched back to v5: https://github.com/remix-run/react-router/issues/10814

**Comment by ptmkenny** [comment]
@aeharding Thanks for posting this.

If anyone else finds important bugs with React Router 5 that have been fixed in React Router 6/7, please feel free to post them in this thread and I will add them to the issue summary.

**Comment by yannisvanlaerhoven3** [comment]
2025 and still no v6 support 😬

**Comment by mleister97** [comment]
What is the current status of this GitHub issue? We're now in 2025, and react-router `7.1.3` has been released. I understand that this might be a challenging task, but **please update us on the current progress** and provide a rough timeline for when we can expect it.

**Comment by nadavhalfon** [comment]
> Hello everyone,
> 
> I wanted to provide an update on this issue — apologies for not doing so sooner. This is our most upvoted and highest-priority issue, so it remains a key focus for us.
> 
> Upgrading React Router has been a highly complex task, which is why it has taken longer than we initially hoped. There have already been two attempts to address this upgrade. The [most recent attempt](https://github.com/ionic-team/ionic-framework/compare/main...sp/react-router-6-take-2) made significant progress, but there are still some unresolved issues.
> 
> Currently, we have an open task underway, and our team is actively investigating and addressing the remaining challenges.
> 
> Thank you for your patience and support as we work towards resolving this.

Hey @brandyscarney,
It’s been 2–3 months since your last update. Could you share any progress on this?
Thanks!

**Comment by nadavhalfon** [comment]
Ok after forking the repo and trying the react router "take-2" branch by @sean-perkins
I can see that there main problem with the react router v6 and above is that it aggressively renders only 1 screen at a time in the DOM.
So the ionic router can't manage animations and gestures.
I don't mind contributing but I'll need some help from the team ✌️

**Comment by ptmkenny** [comment]
For those of you who are interested in using a router other than React Router, I commented on [tanstack router's RFC about SPA mode specifically requesting that support for Ionic be considered](https://github.com/TanStack/router/discussions/3394#discussioncomment-12291664).  Feel free to upvote to ask Tanstack to consider Ionic as a use case.

**Comment by oliveryasuna** [comment]
@nadavhalfon I have opened a discussion on React Router (Remix): https://github.com/remix-run/remix/discussions/10586

**Comment by brandyscarney** [comment]
Hi everyone,

First off, thank you for your continued support and passion for this issue. We know it's been a long-standing request, and we truly appreciate your patience.

I'm sorry this hasn't been resolved yet. Internally, we've been advocating for it and pushing for prioritization with those in charge of the roadmap. Over the past year, though, we've gone through a lot of internal changes. We lost several team members, and the organization shifted focus, which unfortunately meant this work stalled. The last person who had picked it up a few months ago left the company before making much progress.

That said, things are turning around. You may have noticed we've been releasing more consistently again. About a month ago, a newly formed, smaller team officially split off to focus [fully on open source](https://ionic.io/blog/coming-soon-new-updates-features-for-ionics-open-source-projects). We're still in the process of ramping up, but we're making steady progress and are fully focused on our goals.

It's now approved for a couple of team members (already assigned) to begin working on React Router support again in two weeks. It's officially back on the roadmap, and we'll share updates as we go. We're aiming for this to be the final push to get this issue resolved.

Thanks again for sticking with us.

**Comment by kientrant1** [comment]
There are two new security report from using react-router v5 in Ionic. Hope that new Ionic version will be released soon to fix them!

1. https://github.com/advisories/GHSA-cpj6-fhp6-mr6j
2. https://github.com/advisories/GHSA-f46r-rw29-r322

**Comment by aeharding** [comment]
> There are two new security report from using react-router v5 in Ionic. Hope that new Ionic version will be released soon to fix them!
> 
>     1. [GHSA-cpj6-fhp6-mr6j](https://github.com/advisories/GHSA-cpj6-fhp6-mr6j)
> 
>     2. [GHSA-f46r-rw29-r322](https://github.com/advisories/GHSA-f46r-rw29-r322)

From those links:

Affected versions
>= 7.2.0, <= 7.5.1

So Ionic router isn't effected by them

**Comment by ptmkenny** [comment]
@kientrant1 Also note that React Router v5 is NOT end of life and security fixes will still be issued, if applicable.  So although we are all eagerly awaiting support for v7, lack of security updates is not a concern here.

**Comment by ptmkenny** [comment]
@kientrant1 I hadn't seen that before; thanks for sharing.

The relevant links I have from react-router maintainers are the statement [there is no end of life for any react-router version](https://github.com/remix-run/react-router/discussions/9637) (2022) and [there is no official end-of-life policy](https://github.com/remix-run/react-router/issues/11541) (2024; this was new to me since I hadn't searched recently).

So it seems they stealth-deprecated 5.x since I last checked in 2022.

**Comment by playaz87** [comment]
It seems like with the [ViewTransition](https://react.dev/blog/2025/04/23/react-labs-view-transitions-activity-and-more#view-transitions) and [Activity](https://react.dev/blog/2025/04/23/react-labs-view-transitions-activity-and-more#activity) components coming to React this issue should be trivial to solve.

**Comment by nadavhalfon** [comment]
Hey @thetaPC I saw your new react-router-v6 branch, If it's ok we'll be happy know how is it going and if you have a time estimation for the development?
Thank you very much

**Comment by thetaPC** [comment]
We're making great progress on updating the Ionic React Router! Our current focus is on documentation.

We don't have a public release timeline just yet, but we'll definitely make an announcement as soon as it's ready, so keep an eye out!

**Comment by daveycodez** [comment]
> We're making great progress on updating the Ionic React Router! Our current focus is on documentation.
> 
> We don't have a public release timeline just yet, but we'll definitely make an announcement as soon as it's ready, so keep an eye out!

If you guys find a way to support hybrid applications that run as SSR on the web and also static capacitor builds that would be a huge win and I will drop Next.js so fast

**Comment by gnbm** [comment]
Hey everyone 👋🏻

Quick update on React Router support: 
The team has been making great progress, but we'll have a short, temporary pause in active development. 
This is due to planned team absences and vacations, temporarily reducing our capacity.

No worries though, this is just a brief interlude of a couple of weeks. We're still fully invested, and work will resume with renewed focus very soon.

Thanks again for sticking with us, and we'll keep you posted.

**Comment by joebnb** [comment]
> We're making great progress on updating the Ionic React Router! Our current focus is on documentation.
> 
> We don't have a public release timeline just yet, but we'll definitely make an announcement as soon as it's ready, so keep an eye out!

waiting for this release, my project already use v7, if downgrade to v5 seems have alot of works

**Comment by nadavhalfon** [comment]
> Hey everyone 👋🏻
> 
> Quick update on React Router support: The team has been making great progress, but we'll have a short, temporary pause in active development. This is due to planned team absences and vacations, temporarily reducing our capacity.
> 
> No worries though, this is just a brief interlude of a couple of weeks. We're still fully invested, and work will resume with renewed focus very soon.
> 
> Thanks again for sticking with us, and we'll keep you posted.

Hey Thanks for the update!
Has the team resume the work? just asking to be updated :)
I saw that we didn't make much progress in the last 2 months inside the new branch

**Comment by ShaneK** [comment]
Hey! Yep, we've resumed working on it in another branch and are making solid progress!

**Comment by nadavhalfon** [comment]
> Hey! Yep, we've resumed working on it in another branch and are making solid progress!

Thanks for the reply!
Do you mind sharing with us the new branch? or is internal?

**Comment by nadavhalfon** [comment]
Hey, any update? Thanks

**Comment by gnbm** [comment]
Hey everyone👋🏻

Thanks for all the great input on this! We're making a couple of changes to clarify our focus.

First, we're renaming this issue back to `feat: React Router v6 support`. 
This issue was created for v6, and while we appreciate the enthusiasm, our immediate business priority and defined scope for the Ionic Framework v9 release (no ETA at the moment) is to nail down the v6 upgrade.

Second, to keep things organized, we'll be creating a separate GitHub issue for React Router v7 support - https://github.com/ionic-team/ionic-framework/issues/30745.
This lets us focus this thread on v6 while keeping v7 on our radar!

We want to be transparent about our scope and process. As the team is currently deep into planning and work for Ionic Framework v9, our business decision was to define the React Router upgrade to v6 as the scope for this next major release. Managing the upgrade to v6 is a significant piece of work on its own, and focusing on it now allows us to deliver a stable and thoroughly tested foundation.

This doesn't mean we aren't looking ahead! We're committed to supporting the community and modernizing our integrations. 
That's why we're going to rely heavily on your help! 
We'll soon have a shareable version available so you can start testing, providing feedback, and bringing us more use cases. 
This community-driven testing is invaluable for ensuring we deliver the best possible product.

We're excited to get this out to you soon and appreciate your understanding and cooperation as we work towards a v6-compatible integration.

Stay tuned for the shareable version announcement!

**Comment by nadavhalfon** [comment]
> Hey everyone👋🏻
> 
> Thanks for all the great input on this! We're making a couple of changes to clarify our focus.
> 
> First, we're renaming this issue back to `feat: React Router v6 support`. This issue was created for v6, and while we appreciate the enthusiasm, our immediate business priority and defined scope for the Ionic Framework v9 release (no ETA at the moment) is to nail down the v6 upgrade.
> 
> Second, to keep things organized, we'll be creating a separate GitHub issue for React Router v7 support - [#30745](https://github.com/ionic-team/ionic-framework/issues/30745). This lets us focus this thread on v6 while keeping v7 on our radar!
> 
> We want to be transparent about our scope and process. As the team is currently deep into planning and work for Ionic Framework v9, our business decision was to define the React Router upgrade to v6 as the scope for this next major release. Managing the upgrade to v6 is a significant piece of work on its own, and focusing on it now allows us to deliver a stable and thoroughly tested foundation.
> 
> This doesn't mean we aren't looking ahead! We're committed to supporting the community and modernizing our integrations. That's why we're going to rely heavily on your help! We'll soon have a shareable version available so you can start testing, providing feedback, and bringing us more use cases. This community-driven testing is invaluable for ensuring we deliver the best possible product.
> 
> We're excited to get this out to you soon and appreciate your understanding and cooperation as we work towards a v6-compatible integration.
> 
> Stay tuned for the shareable version announcement!

Thanks you very much for the update, all of us are really waiting for this

**Comment by ShaneK** [comment]
Hey everyone!

We greatly appreciate your patience on this issue. I [have a PR](https://github.com/ionic-team/ionic-framework/pull/30831) up that is intended to fully move Ionic Framework's React Router to use React Router 6 now.

If you'd like to try it before v9 releases, there's a dev build, but **⚠️ warning ⚠️** If you're going to use this dev build on an existing react project, you'll need to migrate to React Router 6. Migrating a large project at this point might be a bad idea since this will not release until v9, which will require further migrations and have other breaking changes! I do not currently have migration documentation for this, but it's a [relatively easy process](https://github.com/ionic-team/ionic-framework/pull/30831/commits/a23f55519ed17e8e0601dae57536ddecbb9c3d9f#diff-12ecfc4b2e5e92d598c738a04de9040c7ad728973d44bc78db156963cc0dd5d3) from the migrations I've been doing.

The current dev build is here:
```
8.7.12-dev.11764873961.1fedca46
```

Let me know if you try it and if you have any feedback, please!

**Comment by ptmkenny** [comment]
@ShaneK Thanks for your work on this. I tested the PR and created a new branch to integrate the dev build. My app is fairly complex, about 100 routes with four nested router outlets, state derived from the query string, and hosted from a subdirectory when served as a PWA, all of which required special handling in React Router v5.

### Problem with wrapping routes in IonTabBar (fixed)

I got it working. You can split routes that show the tabbar and that do not show the tabbar like this:

```
      <IonReactRouter basename={appBaseName}>
        <IonRouterOutlet>
          <Route path={routes.onboarding}>
           <RouterOutletOnboarding />
          </Route>
           <Route
                path="*"
                element={
                  <AppTabBar>
                    <IonRouterOutlet>
                      <Route
                        path={routes.login}
                        element={
                          <GuardAnonOnlyRoute>
                            <PageLogin />
                          </GuardAnonOnlyRoute>
                        }
                      />
                  </AppTabBar>
               </IonRouterOutlet>
          </Route>
      </IonRouterOutlet>
    </IonReactRouter>
```

I'm still having the [trouble with slashes described below](https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3624311206), but to get the tabs working, the above structure seems to work well.

### Original Post

I've managed to get the app building and I can access some routes, but I'm having trouble wrapping some routes with IonTabBar (the app only renders the tab bar for some routes).

In react-router 5, I implemented "do not show the tab bar on onboarding pages" like this:

```
      <IonReactRouter basename={appBaseName}>
        <Switch>
          <Route path={routes.onboarding}>
           <RouterOutletOnboarding />
          </Route>
          <AppTabBar>
            <IonRouterOutlet>
              <Route
                exact
               path={login}
               component={<PageLogin />}
              />
          </AppTabBar>
        </IonRouterOutlet>
      </Switch>
    </IonReactRouter>
```

However, in react-router 6, `<Routes>`, which replaces `<Switch>`, will only accept `<Route>` as a child. So it seems the structure needs to use a layout route (https://dev.to/jps27cse/understanding-layout-components-and-react-router-outlet-in-react-3l2e) and look something like this:

```
      <IonReactRouter basename={appBaseName}>
        <Routes>
          <Route path={routes.onboarding}>
           <RouterOutletOnboarding />
          </Route>
          <Route element={<AppTabBar><Outlet /></AppTabBar>}>
            <Route
               path={login}
               component={<PageLogin />}
            />
          </Route>
      </Routes>
    </IonReactRouter>
```

The problem is that in React Router 5, the AppTabBar component renders `children`, but for React Router 6, layout components are supposed to use react-router's `<Outlet>` to render the applicable route, but doing so produces an error: `IonTabs must contain an IonRouterOutlet or an IonTab`

Wrapping it like `<AppTabBar><IonRouterOutlet><Outlet /></IonRouterOutlet></AppTabBar>` doesn't work, either. Do you have a recommendation for how to selectively render the tab bar for some routes and not others? (I can make another issue if you would prefer that.)

**Comment by ptmkenny** [comment]
Issue: The behavior with the presence of a slash at the start of the path differs between react-router's Routes and IonRouterOutlet. For IonRouterOutlet, if the path does not start with a slash, the route does not match and does not get rendered. However, the same route will match when using the Routes component.

I am assuming this is an error because I assume IonRouterOutlet is supposed to mirror the Routes component.

Error: No help tab is rendered
```
      <IonReactRouter>
        <IonRouterOutlet>
          <Route path="help" element={<TabHelp />} />
        </IonRouterOutlet>
      </IonReactRouter>
```

OK: Help tab is rendered
```
      <IonReactRouter>
        <IonRouterOutlet>
          <Route path="/help" element={<TabHelp />} />
        </IonRouterOutlet>
      </IonReactRouter>
```

OK: Help tab is rendered
```
      <IonReactRouter>
        <Routes>
          <Route path="help" element={<TabHelp />} />
        </Routes>
      </IonReactRouter>
```

OK: Help tab is rendered
```
      <IonReactRouter>
        <Routes>
          <Route path="/help" element={<TabHelp />} />
        </Routes>
      </IonReactRouter>
```

The simple example above shows that the route won't match. A related issue is that if you have even one route that doesn't start with a slash, IonRouterOutlet seems to fail to match any route. You can "fix" this by making sure every route has a slash. But Routes doesn't require the slash, so this is very confusing.

**Comment by ShaneK** [comment]
@ptmkenny 

## Tabs
I'm glad you figured out your tab issue! I was going to look into that today, but really happy you figured it out!

## Routes
Thank you so much for reporting this issue! I've a fix for this in a [different branch](https://github.com/ionic-team/ionic-framework/tree/fix/react-router-6-paths) ([pr](https://github.com/ionic-team/ionic-framework/pull/30844)), here's a dev build if you'd like to try it:
```
8.7.12-dev.11765219790.17cbe2e9
```

I've broken this into a different branch to make it easier to review this fix on an individual level

**Comment by ptmkenny** [comment]
## IonBackButton allows navigation between tabs

When used with tabs, `IonBackButton` is supposed to be separate per tab. Clicking the back button should only go back within that tab.

However, with the current react-router ionic dev build, the back button can be used to navigate across tabs.

### Setup

I have tabbed routes set up as follows (`WrapperIonTabs` contains a tab bar that shows three tabs that link to A, B, and C.):

```
              <Route
                path="*"
                element={
                  <WrapperIonTabs>
                    <IonRouterOutlet>
                      <Route
                        path="a"
                        element={
                          <IonPage>
                            <IonHeader>
                              <IonToolbar>
                                <IonBackButton />
                              </IonToolbar>
                            </IonHeader>
                            <IonContent>
                              <p>a</p>
                            </IonContent>
                          </IonPage>
                        }
                      />
                      <Route
                        path="b"
                        element={
                          <IonPage>
                            <IonHeader>
                              <IonToolbar>
                                <IonBackButton />
                              </IonToolbar>
                            </IonHeader>
                            <IonContent>
                              <p>b</p>
                            </IonContent>
                          </IonPage>
                        }
                      />
                      <Route
                        path="c"
                        element={
                          <IonPage>
                            <IonHeader>
                              <IonToolbar>
                                <IonBackButton />
                              </IonToolbar>
                            </IonHeader>
                            <IonContent>
                              <p>c</p>
                            </IonContent>
                          </IonPage>
                        }
                      />
                    </IonRouterOutlet>
                  </WrapperIonTabs>
                }
              />
```

If you click from one tab to another by clicking on the tab bar, and then click the `IonBackButton`, you will go back to the previous tab-- but actually the back button shouldn't do anything here, because each tab should have a separate back history, and there shouldn't be any navigation history for the tab.

**Comment by ShaneK** [comment]
@ptmkenny 

> ## IonBackButton allows navigation between tabs
> 
> [...]

PR to fix that, which includes its own dev build: https://github.com/ionic-team/ionic-framework/pull/30854

**Comment by ptmkenny** [comment]
## Presence of a 404 route causes the base route to be repeated

Somehow, the presence of a 404 route is causing the tab path to be repeated. 

Repro: https://github.com/ptmkenny/ionic-react-router-6-test/tree/router-outlet2

Instructions are in the README.

**Comment by ShaneK** [comment]
> ## Presence of a 404 route causes the base route to be repeated
> 
> [...]

Can you try with this dev build and see if it's better or if you're still experiencing weirdness?

```
8.7.12-dev.11765390930.11e7051a
```

**Comment by ptmkenny** [comment]
> Can you try with this dev build and see if it's better or if you're still experiencing weirdness?

The repeating of `/tab1/tab1` no longer occurs, but the route is not matching. To get the route to match, I have to change `path="tab1/*"` in `/tab1/*` in App.tsx, which might be fixed by https://github.com/ionic-team/ionic-framework/pull/30844 if that isn't already incorporated into this build.

**Comment by ShaneK** [comment]
@ptmkenny I have a PR up with a new dev build that I've verified works with your app now: https://github.com/ionic-team/ionic-framework/pull/30861

**Comment by ptmkenny** [comment]
## `defaultHref` does not work for back button in tab routes

Broken build:

https://github.com/ptmkenny/ionic-react-router-6-test/pull/new/back-button-set

Steps to reproduce:

1. `npm run dev`
2. Go to `http://localhost:5173/tab1`
3. Click `ABC Page (relative)`
4. Click the back button; it will work correctly.
5. Now go to `http://localhost:5173/tab1/abc`
6. The back button will appear (`defaultHref`). However, clicking it does nothing.

**Comment by ptmkenny** [comment]
## When a modal isn't closed properly, IonReactRouter can be set to aria-hidden when it has visible content

When you have two or more IonRouterOutlets nested in a parent IonRouterOutlet, and the user opens a modal and clicks a button in that modal that navigates to another IonRouterOutlet, if the modal is not closed properly, the modal will be closed automatically but the `aria-hidden` on the parent router outlet will remain, which breaks usability and playwright tests. (If the modal is closed properly, then this does not occur, so this may not be a bug. However, in React Router 5, even if the modal is not closed properly, the aria-hidden will not remain set for a parent IonRouterOutlet when a child IonRouterOutlet has visible content.)

Repro:

https://github.com/ptmkenny/ionic-react-router-6-test/pull/new/aria-hidden

Steps to reproduce:

1. `npm run dev`
2. Go to `http://localhost:5173/tab1`
3. Click `Open modal` button
4. Click `Go to onboarding/verify-email` button
5. The parent `IonRouterOutlet` will have aria-hidden set, even though it should not be (it was hidden while the modal was displayed, but since the modal is no longer displayed, it should not be hidden anymore)

**Comment by ptmkenny** [comment]
Hey, I just wanted to report that I now have react-router 6 working in my Ionic app (all automated tests passing) thanks to @ShaneK. You've done a great job with this!

It took awhile to refactor the router outlets I was using and to undo some code that served as workarounds for weird behavior of react-router 5, but now everything is working great. Many thanks!

**Comment by msio** [comment]
Hi @ptmkenny , can you share demo project with react router 6 ?

**Comment by ptmkenny** [comment]
@msio The only demo projects I have are the various repros for bug reports linked to in this thread.

**Comment by benhurley** [comment]
Hi, any timeline on when this will be available?

**Comment by gnbm** [comment]
> Hi, any timeline on when this will be available?

Hi @benhurley 
React Router v6 support will be a crucial part of Ionic Framework v9, since it will be included in our next release.
We're currently planning the next major, and with the information we have, it will be in the second half of the year, but no ETA is defined at the moment.

**Comment by daveycodez** [comment]
I really feel like v6 should be skipped for v7 and TanStack Start.

**Comment by iAMkVIN-S** [comment]
I agree that if the transition from react-router 5 to 6 has already been an open issue since 2021 and with most of the movement on this being in the last few months, there might be some value in pursuing the transition process from [6 to 7](https://reactrouter.com/upgrading/v6) considering it seems like 6 wouldn't see the light of day until at least Q3 of 2026. I fear that at this point sticking to 6 might risk Ionic to still be two major versions behind with the eventual release of [react-router 8](https://github.com/remix-run/react-router/discussions/14468).

However, @daveycodez I'm not sure how much value there would be to migrate the core ionic repository to run from TanStack Start...
- For one, it uses its own routing, which would obviouly nullify any use of react-router. 
- Secondly, similar to nextjs, most of the benefits of TanStack Start such as SSR, Server Functions, etc. don't really apply the same way to the majority of Ionic users who deploy iOS and Android applications since the compiled web assets are already on device.

**Comment by ptmkenny** [comment]
@daveycodez @iAMkVIN-S The decision is to do 6 first and then 7, see the comment in this issue here: https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3434211398 This thread is for the feature request for React Router 6; if you want to comment on 7, please do so on the 7 thread: https://github.com/ionic-team/ionic-framework/issues/30745

There is already a PR in testing to add 6 support, so this is not really the place to discuss the router migration policy. You are welcome to test the 6 PR and report back any bugs you find here.

**Comment by iAMkVIN-S** [comment]
@ptmkenny Agreed! Thank you for the links 🙏

**Comment by ShaneK** [comment]
Hey everyone! I've fixed a *ton* of issues in the [React Router 6 PR](https://github.com/ionic-team/ionic-framework/pull/30831) as we prepare V9. If you'd like to try it out, I've also worked on a [migration guide](https://ionic-docs-git-v9-react-router-ionic1.vercel.app/docs/updating/9-0#react-router) so you can give it a shot!

I'd *strongly* recommend doing this on a new branch if you're going to convert an existing project, since this will be part of v9 which will have some other very major changes.

New dev build:
```
8.8.1-dev.11773432149.19826d0c
```

**Comment by ptmkenny** [comment]
Awesome!

I updated my app that I previously migrated to React Router 6. Route fallbacks are tighter now, so I had to adjust some routes (this is an improvement, as it seems ionic's react router is now more closely following RR6), and now everything is basically working except the `aria-hidden` bug [referenced in this comment](https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3763474151) is still present, and it may even have gotten a little worse because I was able to reproduce it with the new build even when I close the modals correctly.

I [updated the repro](https://github.com/ptmkenny/ionic-react-router-6-test/tree/aria-hidden) to use the new version and I added a [claude code cause analysis](https://github.com/ptmkenny/ionic-react-router-6-test/blob/aria-hidden/router-aria-hidden.md).

**Comment by ShaneK** [comment]
Hey @ptmkenny, thanks for that! I think I fixed that issue in this dev build, if you wanna try it:
```
8.8.1-dev.11774029927.130994f5
```

**Comment by aeharding** [comment]
Just tried it out and seems very promising! I did find one issue, repro here:

https://github.com/aeharding/rr6-back

1. `ionic serve`
2. Fresh load http://localhost:8100/tab1/blah **directly** - NOT the index first
3. Try to click back button, observe does not work

The back button `defaultHref` appears to not be working like in rr5

**Comment by ptmkenny** [comment]
@aeharding Thanks, that looks like a known issue: https://github.com/ionic-team/ionic-framework/issues/24177#issuecomment-3726846690

**Comment by ShaneK** [comment]
Hey, thanks @aeharding and @ptmkenny! I have no idea how I missed this before, I'm sorry! I've fixed this [in my last commit](https://github.com/ionic-team/ionic-framework/pull/30831/commits/c3b959828724daa6739d67bceea6fd7e816651c7) and I have a new dev build here:

```
8.8.1-dev.11774384072.1e807ca8
```

Are there any other reported bugs I missed? I think I got everything else reported either in this thread or the PR itself 🤔 I've also been going through and fixing a ton of issues that were previously reported with our RR5 implementation, so this upgrade will be a massive improvement in many ways!

**Comment by ptmkenny** [comment]
@ShaneK I can confirm `8.8.1-dev.11774384072.1e807ca8` fixes the back bug. As for other reported bugs, is it safe to assume that `8.8.1-dev.11774384072.1e807ca8` is supposed to fix all known bugs? I will re-check the other bugs, but there have been a lot of builds, so I want to make sure I test the right one before reporting stuff that isn't working that has actually already been fixed.

**Comment by ShaneK** [comment]
> @ShaneK I can confirm `8.8.1-dev.11774384072.1e807ca8` fixes the back bug. As for other reported bugs, is it safe to assume that `8.8.1-dev.11774384072.1e807ca8` is supposed to fix all known bugs? I will re-check the other bugs, but there have been a lot of builds, so I want to make sure I test the right one before reporting stuff that isn't working that has actually already been fixed.

Every bug that I know about is fixed in the current build, which is that one. This includes ones reported in this thread, the other thread, and all of the old RR 5 bugs I could dig up. There's also a new feature I haven't documented yet, which allows navigating and resetting the nav stack - useful for things like logging in/out.

So yeah, right now as far as I know there's no reported bugs around.

**Comment by ptmkenny** [comment]
@ShaneK Ok, I went back and checked the bugs I reported, and the current build fixes all of them, awesome!

However, I may have identified a regression, repro here: https://github.com/ptmkenny/ionic-react-router-6-test/tree/404-issue

In my app, some pages are loaded via the network. As described in the repro README, this worked in the previous build, but no longer works in the current build (get a "maximum update depth exceeded" error).

**Comment by ShaneK** [comment]
@ptmkenny Yeah, so that's an issue specifically caused by React 19, unfortunately. I guess suspend in React 19 doesn't unmount components, but instead hides them, causing an infinite loop in the Ionic react router. **_We don't yet support React 19 officially_** because our React output target version from Stencil has some issues with it, so I'm unable to create an actual regression test for this. However, thanks to your reproduction, I was able to fix it by testing it against your code, which will be good to future proof us for when we can support React 19.

The new dev build with the fix for that issue is here:
```
8.8.1-dev.11774636992.1511ce70
```

I've also verified the dev build works in your provided reproduction.

**Comment by ptmkenny** [comment]
@ShaneK  I'm sorry, I didn't realize there wasn't official React 19 support yet. That said, I've been using React 19 since it came out with my app and I haven't had any problems other than the one I just reported.

I tested the new build in my app and the infinite loop is resolved. Thank you so much.

Also, when you're ready for people to test the nav stack reset, I'd be happy to do so; right now, my app just force-reloads the window on logout to get a new nav stack; it would be nice to have a cleaner solution.

**Comment by ShaneK** [comment]
@ptmkenny That's great to hear! You can actually use the stack reset call right now, you just call [navigateRoot from a useIonRouter](https://github.com/ionic-team/ionic-framework/commit/e807ca885e0ef9d676f0ea92e5a99ddc9567a24b). I just don't have it documented yet because I was waiting for the main migration PR to be ready, but now that it's merged I'll be creating the PR for this documentation on Monday. 

It's pretty straight forward though if you want to give it a shot

Edit: Documentation PR is up, [preview link for the new method is here](https://ionic-docs-git-v9-react-utility-functions-ionic1.vercel.app/docs/react/utility-functions#navigateroot)

**Comment by ShaneK** [comment]
Hey! Just letting people who are interested in the dev build know that I just updated major-9.0 from main for the first time in a while and the RR6 branch off of that and created a new dev build. This one is based off the [newly released 8.8.3](https://github.com/ionic-team/ionic-framework/releases/tag/v8.8.3):

```
8.8.4-dev.11775078622.1402ffa2
```

**Comment by ShaneK** [comment]
The final dev build for the RR6 branch is up:
```
8.8.4-dev.11777318673.18d001f6
```

Note, this has changes from major-9.0 which do remove support for React 17 due to a Stencil upgrade, but that's the only breaking change I know of in the branch currently. After this point, the RR6 branch has been merged into major-9.0 and will have any breaking changes that branch incurs, including the upcoming Ionic Modular changes.

**Comment by ptmkenny** [comment]
@ShaneK Thanks for keeping the build updated! I have tested the second-to-last build `8.8.4-dev.11775078622.1402ffa2` pretty extensively and it's working great.  I am also using the stack reset, which surfaced several bugs in my app and allowed me to improve logging a lot since logout no longer requires a reload.

I started to test the final build `8.8.4-dev.11777318673.18d001f6` but there seem to be a lot of stencil changes related to how components are constructed. These broke my playwright tests (things like name attribute not being set for checkboxes anymore, and host attributes no longer reflecting component state or labels after state changes).  I imagine the stencil stuff is still in development and not related to this PR, but just a heads up for anyone trying to test this now.

**Comment by Huaian666** [comment]
about the react part — +1 on this. would save so much boilerplate and make the API feel more complete

**Comment by ShaneK** [comment]
@Huaian666 I'm not sure what react part you're referring to? Sorry?

@ptmkenny That was actually a bug in Stencil that was surfacing all of those properties by accident. It wasn't an intentional feature and was fixed in Stencil [a very long time ago](https://github.com/stenciljs/output-targets/releases/tag/%40stencil%2Freact-output-target%400.6.0) by making props require reflect to show up, but we were stuck on an old version because of breaking changes so it still existed in Ionic Framework's React package

**Comment by ptmkenny** [comment]
@ShaneK I see. I started to refactor my playwright tests, and I came across a major issue: due to the stencil change, form validation classes applied by a third-party form manager are broken.  Since this is not related to React Router and this already got committed, I opened a new issue here: https://github.com/ionic-team/ionic-framework/issues/31237

**Comment by ShaneK** [comment]
Hey everyone! Thank you so much for your patience! This issue is now resolved, as RR6 support is in Ionic Framework v9, [out today!](https://ionic.io/blog/announcing-ionic-framework-9)

**Comment by ptmkenny** [comment]
As a final update, today my app (ionic framework 9.0.3, capacitor 9 alpha5, and react router 6) was approved by the App Store review team!  Performance is much better and I was able to eliminate several weird bugs compared to the React Router 5 version. Many thanks to @ShaneK in particular as well as everyone else who helped to get this done over the past few years.
